util/ranger/points.go

```diff
@@ -421,11 +422,27 @@ func (r *builder) buildFromNot(expr *expression.ScalarFunction) []point {
 	case ast.IsFalsity:
 		return r.buildFromIsFalse(expr, 1)
 	case ast.In:
-		rangePoints, hasNull := r.buildFromIn(expr)
+		var unsignedInt bool
+		if x, ok := expr.GetArgs()[0].(*expression.Column); ok {
+			unsignedInt = mysql.HasUnsignedFlag(x.RetType.Flag) && mysql.IsIntegerType(x.RetType.Tp)
+		}
+		inPoints, hasNull := r.buildFromIn(expr)
 		if hasNull {
 			return nil
 		}
-		retRangePoints := make([]point, 0, len(rangePoints)+2)
+		var rangePoints []point
+		// negative ranges can be directly ignored for unsigned int columns.
+		if unsignedInt {
+			rangePoints = make([]point, 0, 2+len(inPoints))
```

Comment:
Since the `rangePoints` is in order, the negative one must be the head of the slice.
So we can just get the position and use `rangePoints = rangePoints[nonNegativePos:]`

Reply:
ok